packages/dashboard/src/components/DashboardGridLayout/DashboardGridLayoutWithAddCardPanel/DashboardGridLayoutWithAddCardPanel.tsx

```diff
@@ -0,0 +1,200 @@
+      // large is 2 scale high and wide
+      // mediumTall is 1 scale wide and 2 scales high
+      if (individualItemLayout.h === 1) {
+        cardSize = CardSize.section;
+      } else if (key === '28') {
+        cardSize = CardSize.large;
+      } else if (key === '18') {
+        cardSize = CardSize.mediumTall;
+      } else if (key === '24') {
+        cardSize = CardSize.mediumWide;
+      }
+      const itemLayout: IDashboardCardLayout = {
+        i: individualItemLayout.i!,
+        x: individualItemLayout.x,
+        y: individualItemLayout.y,
+        size: cardSize
+      };
+      newLayout.lg!.push(itemLayout);
+    });
+    if (newLayout !== this.state.layout) {
+      if (this.props.onLayoutChange) {
+        this.props.onLayoutChange(newLayout);
+      }
+      this.setState({
+        layout: newLayout
+      });
+    }
+  };
+
+  private _addCard = (cardId: string): void => {
+    const addCardPanelCards = this.state.cardsForAddCardPanel;
+    let cardIndex: number = -1;
+    const layout: DashboardGridBreakpointLayouts = this.state.layout;
+    // find the card selected in the list of cards in add card panel
+    addCardPanelCards.map((card: IDGLCard, index: number) => {
+      if (card.id === cardId) {
+        cardIndex = index;
+        // calculate new layout for the selected card from add card panel
+        const calculatedLayout: { x: number; y: number } = this._calculateNextCardPostion(card.cardSize);
+        const cardLayout: IDashboardCardLayout = { i: card.id, x: calculatedLayout.x, y: calculatedLayout.y, size: card.cardSize };
+        layout.lg!.push(cardLayout);
+      }
+    });
+    if (cardIndex !== -1) {
+      // remove the selected card from the add card panel and add it to the list of cards that are to be show in layout
+      const cardSelected = addCardPanelCards.splice(cardIndex, 1);
+      const layoutCards: IDGLCard[] = this.state.cardsForLayout;
+      layoutCards.push(cardSelected[0]);
+      this.setState({
+        cardsForAddCardPanel: addCardPanelCards,
+        cardsForLayout: layoutCards,
+        layout: layout
+      });
+      // scroll to the card that was added to the layout
+      setTimeout(() => {
+        if (document.getElementById(cardId + 'dglCard')) {
+          document.getElementById(cardId + 'dglCard')!.scrollIntoView({ behavior: 'smooth' });
+          const css = getClassNames(getCardStyles!);
+          document.getElementById(cardId + 'dglCard')!.classList.add(css.fadeIn);
+        }
+      }, 100);
+    }
+  };
+
+  // calculate the position for the selected card from add card panel
+  private _calculateNextCardPostion = (cardSize: CardSize) => {
+    let newCardXPosition: number = 0;
+    let newCardYPosition: number = 0;
+    let maxYX: number = 0;
+    let lastButOneYX: number = 0;
+    const layoutState = this.state.layout;
+    const lastElement: undefined | IDashboardCardLayout = layoutState.lg![layoutState.lg!.length - 1];
```

Comment:
>layoutState.lg!.length - 1 [](start = 74, length = 26)

how do you the layout is > 0. What if there are no cards to begin with

Reply:
If there are no cards to begin with undefined is returned, accounted for that condition in the next lines.